### PR TITLE
JNT-149

### PR DESCRIPTION
**Pull request by abhilabh**
Fix for JNT-149 to restrict user to allocate multiple RJIL IP to instances.

### CODE REVIEW

File: ec2api/exception.py [hunk 1]
```diff
@@ -281,6 +281,9 @@ class CannotDelete(EC2IncorrectStateException):
 class ResourceAlreadyAssociated(EC2IncorrectStateException):
     ec2_code = 'Resource.AlreadyAssociated'
 
+class AllreadyRjilIPAssociated(EC2IncorrectStateException):
```
Comment (shashankchakelam) [inline ec2api/exception.py]: Can you fix typo: "Already"

File: ec2api/api/address.py [hunk 2]
```diff
@@ -306,6 +306,14 @@ def associate_address(self, context, public_ip=None, instance_id=None,
             for eni in db_api.get_items(context, 'eni'):
                 if instance_id and eni.get('instance_id') == instance_id:
                     instance_network_interfaces.append(eni)
+                    #Bellow changes done for JNT-149.It will restrict user to allocate more than one
```
Comment (shashankchakelam) [inline ec2api/api/address.py:309]: nit: You mean "only one" not "more than one", right?

File: ec2api/api/address.py [hunk 3]
```diff
@@ -306,6 +306,14 @@ def associate_address(self, context, public_ip=None, instance_id=None,
...
+                    # RJIL IP for an instances
+                    eni_id= eni['id']  #store network interface id
+                    for epi in db_api.get_items(context, 'eipalloc'):
+                       #If network interface Id will be present in eipalloc data then it will throw the exception
+                       if 'network_interface_id' in  epi:
+                        if eni_id == epi['network_interface_id']:
```
Comment (shashankchakelam) [inline ec2api/api/address.py]: Is this check necessary? In case these two are not equal, we would still want to throw the same exception, right?

**Comment by shashankchakelam** [comment]
General comment: can we add unit test cases for these changes (https://github.com/JioCloudVPC/ec2-api/blob/stable/kilo/ec2api/tests/unit/test_security_group.py)

File: ec2api/api/address.py (hunk 2 above)
Comment (abhilabh) [inline ec2api/api/address.py:309]: yes shashank only one RJIL IP per instances.

File: ec2api/api/address.py (hunk 3 above)
Comment (basilsaji-ril) [inline ec2api/api/address.py]: Hi Shashank,

This is the check that confirms if it has a floating IP attached or not. Basically, the flow is

1) Retrieve the eni object by checking the instance id.
2) Check if the above retrieved eni has an associated FIP by checking if it is linked to a FIP resource(eipalloc). Only if there is a link, we can gather that a fip resource has been associated. The above check accomplishes that and is hence needed.

Abhishek,

Is the alignment proper? The diff is showing it a bit off.

**Comment by hmdesai89** [comment]
Looks ok to me and looks like it will work just fine. However personally I feel keeping a flag in 'eni' will do a better job than iterating over 'eips' everytime we want to attach a floating IP

**Pull request merged**